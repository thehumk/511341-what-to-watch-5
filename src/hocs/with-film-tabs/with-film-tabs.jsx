import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {propsForFilms, propsForComments} from '../../utils/prop-types';
import {fetchCommentsFilm} from '../../store/api-actions';

const TypeTabs = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`,
};

export const withFilmTabs = (Component) => {
  class WithFilmTabs extends React.PureComponent {
    constructor(props) {
      super(props);

      this.film = props.film;

      this.state = {
        tab: TypeTabs.OVERVIEW,
      };

      this.tabsClickHandler = this.tabsClickHandler.bind(this);
    }

    componentDidMount() {
      this.props.getFilmComments(this.film.id);
    }

    componentDidUpdate() {
      if (this.film.id !== this.props.film.id) {
        this.setState({tab: TypeTabs.OVERVIEW});
        this.film = this.props.film;
        this.props.getFilmComments(this.film.id);
      }
    }

    tabsClickHandler(evt, type) {
      evt.preventDefault();

      if (this.state.tab === type) {
        return;
      }

      this.setState({tab: type});
    }

    render() {
      const {filmComments} = this.props;

      return (
        <Component
          {...this.props}
          film={this.film}
          filmComments={filmComments}
          tab={this.state.tab}
          tabsClickHandler={this.tabsClickHandler}
        />
      );
    }
  }

  WithFilmTabs.propTypes = {
    film: propsForFilms,
    filmComments: PropTypes.arrayOf(propsForComments).isRequired,
    getFilmComments: PropTypes.func.isRequired,
  };

  const mapStateToProps = ({COMMENTS}) => ({
    filmComments: COMMENTS.comments,
  });

  const mapDispatchToProps = (dispatch) => ({
    getFilmComments(id) {
      dispatch(fetchCommentsFilm(id));
    },
  });

  return connect(mapStateToProps, mapDispatchToProps)(WithFilmTabs);
};
