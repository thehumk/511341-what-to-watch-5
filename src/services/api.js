import axios from 'axios';

const BASE_URL = `https://5.react.pages.academy/wtw`;
const REQUEST_TIMEOUT = 5000;

export const HttpCode = {
  UNAUTHORIZED: 401,
};

export const createAPI = (onUnauthorized) => {
  const api = axios.create({
    baseURL: BASE_URL,
    timeout: REQUEST_TIMEOUT,
    withCredentials: true,
  });

  const onSuccess = (response) => response;

  const onFail = (err) => {
    const {response} = err;

    if (response.status === HttpCode.UNAUTHORIZED) {
      onUnauthorized();

      throw err;
    }
  };

  api.interceptors.response.use(onSuccess, onFail);

  return api;
};
