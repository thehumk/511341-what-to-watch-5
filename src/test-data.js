export const noop = () => {};

export const MATCH = {params: {id: `1`}};

export const mockComments = [
  {
    comment: `123`,
    date: `2020-10-29T13:38:44.769Z`,
    id: 1,
    rating: 4.6,
    user: {
      id: 12,
      name: `Isaac`
    },
  },
  {
    comment: `1234`,
    date: `2020-10-29T13:38:45.769Z`,
    id: 2,
    rating: 4.5,
    user: {
      id: 13,
      name: `Isa`
    },
  }
];

/* eslint-disable */

export const mockFilms = [
  {
    background_color: `#73B39A`,
    background_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    description: `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
    director: `Nicolas Winding Refn`,
    genre: `Action`,
    id: 1,
    is_favorite: true,
    name: `Bronson`,
    poster_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    preview_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: 3.6,
    released: 2008,
    run_time: 92,
    scores_count: 109661,
    starring: [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
    video_link: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  },
  {
    background_color: `#73B39A`,
    background_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    description: `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
    director: `Nicolas Winding Refn`,
    genre: `Drama`,
    id: 2,
    is_favorite: false,
    name: `Bronson`,
    poster_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    preview_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: 3.6,
    released: 2008,
    run_time: 92,
    scores_count: 109661,
    starring: [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
    video_link: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  },
];

/* eslint-enable */

export const state = {
  DATA: {
    films: mockFilms,
    promoFilm: mockFilms[0],
    genresList: [`All`, `Action`, `Drama`],
    isApplicationReady: false,
  },
  FILMS_STATUS: {
    activeGenre: `All`,
    renderFilmsCount: 8,
  },
  USER: {
    authorizationStatus: `NO_AUTH`,
    user: null,
  },
  COMMENTS: {
    comments: mockComments,
  },
};
