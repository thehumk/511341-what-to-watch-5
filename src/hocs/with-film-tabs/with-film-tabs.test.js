import React from 'react';
import {Provider} from 'react-redux';
import configureMockStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import thunk from 'redux-thunk';
import {createAPI} from '../../services/api';
import {withFilmTabs} from './with-film-tabs';
import {noop, mockFilms, comments, state} from '../../test-data';

const api = createAPI(noop);

const mockStore = configureMockStore([thunk.withExtraArgument(api)])(state);

const MockComponent = () => {
  return <div></div>;
};

const MockComponentContainer = withFilmTabs(MockComponent);

it(`Should WithFilmTabs render correctly`, () => {
  const tree = renderer.create(
      <Provider store={mockStore}>
        <MockComponentContainer
          film={mockFilms[0]}
          filmComments={comments}
          getFilmComments={noop}
        />
      </Provider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
