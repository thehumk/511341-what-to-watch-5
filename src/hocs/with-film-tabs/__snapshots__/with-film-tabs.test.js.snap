// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should WithFilmTabs render correctly 1`] = `<div />`;
